sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -187,11 +187,11 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
   /**
    * Prepare a SparkPlan for execution. It's idempotent.
    */
-  final def prepare(): Unit = {
+  final def prepare(): Unit = synchronized {
```

Comment:
This looks dangerous. Child's `doPrepare` is wrapped by `parent's synchronized` without any comment mentioning it. If Child's `doPrepare` starts a new thread and calls `prepare` in it, we may have some dead-lock race.